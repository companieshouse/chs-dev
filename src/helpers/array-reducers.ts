export const collect = <U, T>(fn: (value: T) => U[]): (result: U[], value: T) => U[] => {
    return (values: U[], value: T): U[] => {
        values.push(...fn(value));
        return values;
    };
};

export const deduplicate = (uniqueValues: string[], value: string): string[] => {
    if (!uniqueValues.includes(value)) {
        uniqueValues.push(value);
    }
    return uniqueValues;
};
