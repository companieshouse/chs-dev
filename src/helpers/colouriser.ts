import { cyan, gray, green, red, yellow } from "ansis";

type ColourMap = [RegExp, (str: string) => string][];

const statusColourMap: ColourMap = [
    [/Exited\s\([1-9]|\d{2,}\)/, red],
    [/Exited\s\(0\)/, yellow],
    [/Up[^()]+$/, green],
    [/Up.+\(health:\sstarting./, cyan],
    [/Up.+\(healthy./, green.bold],
    [/Up.+\(unhealthy/, red.bold],
    [/Not\srunning/, gray]
];

const runStatusColourMap: ColourMap = [
    [/Started/, green],
    [/Healthy/, green.bold],
    [/Stopped/, yellow]
];

const stopStatusColourMap: ColourMap = [
    [/Stopped/, yellow],
    [/Removed/, red]
];

const colourise = (colourMap: ColourMap) => (str: string) => {
    const colour = colourMap
        .find(([regex, _]) => regex.test(str));

    if (colour) {
        return colour[1](str);
    } else {
        return str;
    }
};

type SimpleColour = "cyan" | "red" | "bold-red" | "green" | "bold-green" | "yellow" | "gray" | "grey"

export const simpleColouriser = (str: string, colour: SimpleColour) => {
    let colouriseFn: (str: string) => string;

    switch (colour) {
    case "cyan":
        colouriseFn = cyan;
        break;
    case "red":
        colouriseFn = red;
        break;
    case "bold-red":
        colouriseFn = red.bold;
        break;
    case "green":
        colouriseFn = green;
        break;
    case "bold-green":
        colouriseFn = green.bold;
        break;
    case "yellow":
        colouriseFn = yellow;
        break;
    case "gray":
    case "grey":
        colouriseFn = gray;
        break;
    default:
        colouriseFn = (strToColourise: string) => strToColourise;
        break;
    }

    return colouriseFn(str);
};

export const statusColouriser = colourise(statusColourMap);
export const runStatusColouriser = colourise(runStatusColourMap);
export const stopStatusColouriser = colourise(stopStatusColourMap);
