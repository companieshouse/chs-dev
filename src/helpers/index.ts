/**
 * Checks if a given value is an object and not an array.
 *
 * @param {unknown} T - The value to check.
 * @returns {boolean} - Returns `true` if the value is an object and not an array, otherwise `false`.
 */
export const isObjectFormat = (T: unknown): boolean => {
    return typeof T === "object" && !Array.isArray(T) && T !== null;
};

/**
 * Determines if an object is empty.
 *
 * @param {Record<string, any>} obj - The object to check.
 * @returns {boolean} - Returns `true` if the object has no own enumerable properties, otherwise `false`.
 */
export const isEmptyObject = (obj: Record<string, any>): boolean => {
    return Object.keys(obj).length === 0;
};

/**
 * Checks if there is at least one common element between two arrays.
 *
 * @param {unknown[]} array1 - The first array to compare.
 * @param {unknown[]} array2 - The second array to compare.
 * @returns {boolean} - Returns `true` if any element in `array1` exists in `array2`, otherwise `false`.
 */
export const matchExistInArrays = (array1: unknown[], array2: unknown[]): boolean => {
    return array1.some(item => array2.includes(item));
};

/**
 * Finds and returns elements from the parentArray that do not exist in the childArray.
 *
 * @param {unknown[]} parentArray - The source array to check.
 * @param {unknown[]} childArray - The reference array to compare against.
 * @returns {any[]} - An array containing elements from `parentArray` that are not present in `childArray`.
 */
export const findUniqueItemsInParentArray = (parentArray: unknown[], childArray: unknown[]): any[] => {
    const childSet = new Set(childArray);
    return parentArray.filter(item => !childSet.has(item));
};
