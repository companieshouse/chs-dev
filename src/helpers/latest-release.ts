import { satisfies } from "semver";
import { getPersonalAccessToken } from "./github.js";

const githubReleasesApiUrl = "https://api.github.com/repos/companieshouse/chs-dev/releases";
const latestReleaseUrl = `${githubReleasesApiUrl}/latest`;

/**
 * fetches the latest version name from GitHub
 * @returns latest version name
 */
export const getLatestReleaseVersion: (gitRepoUrl?: string) => Promise<string> =
    async (gitRepoUrl = latestReleaseUrl) => {
        const parameters = await githubRequestParameters();
        const response = await fetch(gitRepoUrl, parameters);

        return (await response.json()).name;
    };

/**
 * Finds out the latest version which matches the specification specified.
 * @param versionSpec specification of the version required
 * @returns latest version or undefined if there were no matches
 */
export const getLatestReleaseSatisfying = async (versionSpec: string): Promise<string | undefined> => {
    const githubVersionsIterable = githubReleaseGenerator();

    let githubRelease: IteratorResult<string, undefined>;

    do {
        githubRelease = await githubVersionsIterable.next();

        if (githubRelease.value && satisfies(githubRelease.value, versionSpec)) {
            return githubRelease.value;
        }
    } while (!githubRelease.done);

    return undefined;
};

/**
 * Generator function which will generate/yield version names handling Github API
 * pagination
 */
async function * githubReleaseGenerator (): AsyncGenerator<string, undefined, undefined> {
    // Start with the all releases link
    let releasesLink: string | undefined = githubReleasesApiUrl;

    while (releasesLink) {
        // Fetch page of releases
        const parameters = await githubRequestParameters();
        const releasesResponse = await fetch(releasesLink, parameters);

        // Load body and link header
        const body = await releasesResponse.json();

        const linkHeader = releasesResponse.headers.get("link");

        // Yield release names held in page of results
        for (const release of body) {
            yield release.name;
        }

        // When there is no link header then there was only one page of results
        if (linkHeader === null) {
            releasesLink = undefined;
        } else {
            // Look for next link in the Link header
            const nextLink = /<([^>]+)>;\s+rel="next"/;

            const nextLinkMatches = linkHeader.match(nextLink);

            // If there is next link use this as the next url otherwise break
            if (nextLinkMatches) {
                releasesLink = nextLinkMatches[1];
            } else {
                releasesLink = undefined;
            }
        }
    }
}

const githubRequestParameters = async () => {
    const parameters: {
        headers?: Record<string, string>
    } = {};

    const githubApiToken = await getPersonalAccessToken();

    if (githubApiToken) {
        parameters.headers = {
            Authorization: `Bearer ${githubApiToken}`
        };
    }

    return parameters;
};
