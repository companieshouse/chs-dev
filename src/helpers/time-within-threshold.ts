import { readFileSync } from "fs";

export enum ThresholdUnit {
    MINUTES = 1000 * 60,
    HOURS = 1000 * 60 * 60,
    DAYS = 1000 * 60 * 60 * 24
}

/**
 * Determines whether the ISO date stored within the file date supplied is
 * within the supplied threshold of the date within file `comparisonFile`
 * @param comparisonFile: string - path to file containing ISO date to compare
 *              the executionTime to
 * @param executionTime: Date | number - value to compare to the value in file
 *              which must be within the defined threshold
 * @param thresholdValue: number - number of units which define the acceptable
 *              theshold
 * @param thresholdUnit: ThresholdUnit - The unit of the threshold being tested
 * @returns boolean indicating whether exectutionTime is within the threshold
 * supplied
 */
export const timeWithinThreshold = (
    comparisonFile: string,
    executionTime: Date | number,
    thresholdValue: number,
    thresholdUnit: ThresholdUnit
) => {
    let withinThreshold: boolean = true;

    const lastRun = readFileSync(comparisonFile).toString("utf8");

    if (executionTime instanceof Date) {
        executionTime = executionTime.getTime();
    }

    const difference = executionTime - Date.parse(lastRun);

    const numberOfUnitsLapsed = Math.floor(
        difference / thresholdUnit
    );

    if (numberOfUnitsLapsed >= thresholdValue) {
        withinThreshold = false;
    }

    return withinThreshold;
};
