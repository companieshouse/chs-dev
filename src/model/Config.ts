/**
 * Represents project configuration defining project related properties for
 * customising the behaviours of chs-dev
 */
export type Config = {

    /**
     * name to refer to the project/environment
     */
    readonly projectName: string;

    /**
     * path to the project repository
     */
    readonly projectPath: string;

    /**
     * path to the chs-dev local repository
     */
    readonly chsDevPath?: string;

    /**
     * path to the chs-dev data
     */
    readonly chsDevDataDir?: string;

    /**
     * additional environment variables for running the environment. Values can
     * contain 'file://' prefixes will try to read file in and set as value in
     * its place.
     */
    readonly env: Record<string, string>;

    /**
     * when supplied defines the threshold for number of hours after which to
     * attempt ecr login
     */
    readonly performEcrLoginHoursThreshold?: number;

    /**
     * specifies the supported version of the CLI by the project
     */
    readonly versionSpecification?: string;

}

export default Config;
