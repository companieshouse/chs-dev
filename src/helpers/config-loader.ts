import { existsSync, readFileSync } from "fs";
import { basename, join } from "path";
import yaml from "yaml";
import Config from "../model/Config.js";
import Constants from "../model/Constants.js";
import { getDockerSettingsFilePath } from "./docker-settings-config.js";

const fileVarRegExp = /^file:\/\/(.+)$/;

/**
 * Loads the configuration from the project root. When project does not contain
 * configuration then returns an empty configuration object.
 * @returns Project Config
 */
export const load: () => Config = () => {
    const projectPath = process.env.CHS_DEV_PROJECT || process.cwd();
    const confFile = join(projectPath, "chs-dev/config.yaml");
    const dockerSettingsPath:string = getDockerSettingsFilePath();

    let config: Partial<Config> = {};

    if (!existsSync(confFile)) {
        config = {
            env: {}
        };

    } else {
        const rawConfiguration = readFileSync(join(confFile)).toString("utf8");

        const projectConfiguration = yaml.parse(rawConfiguration);

        config = {
            env: parseEnv(projectConfiguration.env) || {},
            performEcrLoginHoursThreshold: projectConfiguration.authed_repositories && projectConfiguration.ecr_login_threshold_hours ? projectConfiguration.ecr_login_threshold_hours : undefined,
            versionSpecification: projectConfiguration.version
        };
    }

    if (!config.performEcrLoginHoursThreshold) {
        config = {
            ...config,
            performEcrLoginHoursThreshold: Constants.DEFAULT_PERFORM_ECR_LOGIN_HOURS_THRESHOLD
        };
    }

    return {
        ...config,
        projectPath,
        projectName: basename(projectPath),
        dockerSettingsPath
    } as Config;
};

const parseEnv = (env: Record<string, string> | undefined) => {
    if (!env) {
        return undefined;
    }

    return Object.keys(env)
        .map(key => {
            const value = env[key];

            const fileMatches = fileVarRegExp.exec(value);

            if (fileMatches) {
                const file = fileMatches[1];

                if (file && existsSync(file)) {
                    return [key, readFileSync(file).toString("utf8")];
                }
            }

            return [key, value];
        })
        .reduce((previous, next) => {
            const [key, value] = next;

            previous[key] = value;

            return previous;
        }, {} as Record<string, string>);
};

export default load;
