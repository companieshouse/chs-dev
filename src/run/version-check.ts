import { existsSync, mkdirSync, readFileSync, writeFileSync } from "fs";
import Config from "../model/Config.js";
import { diff, satisfies } from "semver";
import { getLatestReleaseVersion } from "../helpers/latest-release.js";
import { join } from "path";
import { ThresholdUnit, timeWithinThreshold } from "../helpers/time-within-threshold.js";

type VersionCheckLogger = {
    warn: (msg: string) => void;
    info: (msg: string) => void;
}

type CreateArguments = {
    dataDirectory: string;
    runThresholdDays: number;
    logger: VersionCheckLogger;
    config: Config;
}

/**
 * A class which checks the version of the project and logs out any
 * inconsistencies to latest version or project version
 */
export class VersionCheck {

    // eslint-disable-next-line no-useless-constructor
    private constructor (private readonly dataDirectory: string,
        private readonly runThresholdDays: number,
        private readonly logger: VersionCheckLogger,
        private readonly config: Config
    ) {

    }

    /**
     * Runs version check checking the supplied version is approprate for the
     * project and also whether a more recent version is available
     * @param applicationVersion version to check
     * @returns Promise representing outcome - resolves to undefined when
     * successful and rejects when there is an error
     */
    async run (applicationVersion: string) {
        const isEnvVarSet = (envVarName: string) => envVarName in process.env;

        if (!isEnvVarSet("CHS_DEV_NO_PROJECT_VERSION_MISMATCH_WARNING") &&
            !this.isSuitableVersionForProject(applicationVersion)) {

            this.logUnsuitableVersion(applicationVersion, this.config.versionSpecification as string);

            return;
        }

        const runTimeFile = join(this.dataDirectory, "last-version-run-time");
        const executionTime = Date.now();

        if (!timeWithinThreshold(runTimeFile, executionTime, this.runThresholdDays, ThresholdUnit.DAYS)) {
            await this.checkCurrentVersionOfCliIsLatest(applicationVersion);

            writeFileSync(
                runTimeFile,
                new Date(executionTime).toISOString()
            );
        }

    }

    /**
     * Creates a VersionCheck object ensuring the dataDir is present when it is
     * absent
     * @param argumentsObject - containing the properties to create the
     * VersionCheck object
     * @returns new VersionCheck object
     */
    static create (
        {
            dataDirectory,
            runThresholdDays,
            logger,
            config
        }: CreateArguments
    ): VersionCheck {
        if (!existsSync(dataDirectory)) {
            mkdirSync(dataDirectory, { recursive: true });
        }

        return new VersionCheck(dataDirectory, runThresholdDays, logger, config);
    }

    private logUnsuitableVersion (actualVersion: string, requiredVersionSpecification: string) {
        const sepLine = "-".repeat(80);
        this.logger.warn(
            `WARNING: this version: ${actualVersion} does not meet the project requirements: ${requiredVersionSpecification}

Run:

\tchs-dev sync

to update chs-dev to a suitable version
${sepLine}

`
        );
    };

    private isSuitableVersionForProject (applicationVersion: string): boolean {
        return !this.config.versionSpecification || satisfies(applicationVersion, this.config.versionSpecification); ;
    }

    private async checkCurrentVersionOfCliIsLatest (applicationVersion) {
        const latestVersion = await getLatestReleaseVersion();

        const semverDifference = diff(latestVersion, applicationVersion);

        if (semverDifference !== null) {
            this.logVersionDifference(latestVersion, applicationVersion, semverDifference);
        }
    };

    private logVersionDifference (latestVersion: string, currentVersion: string, semverDifference: string) {
        const versionDifference: string = semverDifference === "major" || semverDifference === "minor"
            ? ` ${semverDifference}`
            : "";

        const sepLine = "=".repeat(80);
        const versionOutOfDateMessage = `ðŸ“£ There is a newer${versionDifference} version (${latestVersion}) available (current version: ${currentVersion})

Run:

\tchs-dev sync

to update to latest version
`;

        this.logger.info(
            `${sepLine}\n\n${versionOutOfDateMessage}\n\n${sepLine}`
        );
    };

}
