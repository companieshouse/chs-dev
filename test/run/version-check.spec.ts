import { expect, jest } from "@jest/globals";
import Config from "../../src/model/Config";
import { VersionCheck } from "../../src/run/version-check";
import { getLatestReleaseVersion as getLatestReleaseVersionMock } from "../../src/helpers/latest-release";
import { timeWithinThreshold as timeWithinThresholdMock } from "../../src/helpers/time-within-threshold";
import fs from "fs";
import { join } from "path";

jest.mock("../../src/helpers/latest-release");
jest.mock("../../src/helpers/time-within-threshold");

describe("VersionCheck", () => {

    const projectConfig: Config = {
        env: {},
        projectPath: "./docker-project",
        projectName: "docker-project",
        versionSpecification: ">=1.0.0 <2.0.0"
    };
    const loggerMock = {
        warn: jest.fn(),
        info: jest.fn()
    };
    const runThresholdDays = 14;

    const existsSyncSpy = jest.spyOn(fs, "existsSync");
    const mkdirSyncSpy = jest.spyOn(fs, "mkdirSync");
    const writeFileSyncSpy = jest.spyOn(fs, "writeFileSync");
    const readFileSyncSpy = jest.spyOn(fs, "readFileSync");

    const dataDir: string = "data/";

    const version = "1.1.1";

    const differentVersionTestCases = [
        ["1.1.2", `ðŸ“£ There is a newer version (1.1.2) available (current version: ${version})`],
        ["1.2.0", `ðŸ“£ There is a newer minor version (1.2.0) available (current version: ${version})`],
        ["1.2.1", `ðŸ“£ There is a newer minor version (1.2.1) available (current version: ${version})`],
        ["2.0.0", `ðŸ“£ There is a newer major version (2.0.0) available (current version: ${version})`],
        ["2.0.1", `ðŸ“£ There is a newer major version (2.0.1) available (current version: ${version})`],
        ["2.1.0", `ðŸ“£ There is a newer major version (2.1.0) available (current version: ${version})`]
    ];

    describe("create", () => {
        beforeEach(() => {
            jest.resetAllMocks();

            mkdirSyncSpy.mockImplementation((_, __) => {
                return undefined;
            });
        });

        it("checks if data directory exists", () => {
            VersionCheck.create({
                dataDirectory: dataDir,
                logger: loggerMock,
                config: projectConfig,
                runThresholdDays
            });

            expect(existsSyncSpy).toHaveBeenCalledWith(dataDir);
        });

        it("creates data directory when does not exist already", () => {
            existsSyncSpy.mockReturnValue(false);

            VersionCheck.create({
                dataDirectory: dataDir,
                logger: loggerMock,
                config: projectConfig,
                runThresholdDays
            });

            expect(mkdirSyncSpy).toHaveBeenCalledWith(dataDir, { recursive: true });
        });

        it("does not create data directory when already present", () => {
            existsSyncSpy.mockReturnValue(true);

            VersionCheck.create({
                dataDirectory: dataDir,
                logger: loggerMock,
                config: projectConfig,
                runThresholdDays
            });

            expect(mkdirSyncSpy).not.toHaveBeenCalled();
        });

        it("returns a VersionCheck object", () => {

            existsSyncSpy.mockReturnValue(true);

            const result = VersionCheck.create({
                dataDirectory: dataDir,
                logger: loggerMock,
                config: projectConfig,
                runThresholdDays
            });

            expect(result).toBeInstanceOf(VersionCheck);
        });
    });

    describe("run", () => {
        let versionCheck: VersionCheck;

        describe("first run through", () => {
            beforeEach(() => {
                jest.resetAllMocks();

                existsSyncSpy.mockReturnValueOnce(true);

                versionCheck = VersionCheck.create({
                    dataDirectory: dataDir,
                    logger: loggerMock,
                    config: projectConfig,
                    runThresholdDays
                });

                // @ts-expect-error
                getLatestReleaseVersionMock.mockResolvedValue(version);

                delete process.env.CHS_DEV_NO_PROJECT_VERSION_MISMATCH_WARNING;

                writeFileSyncSpy.mockImplementation((_, __) => { });

                existsSyncSpy.mockReturnValue(false);
            });

            it("fetches the latest version", async () => {

                await versionCheck.run(version);

                expect(getLatestReleaseVersionMock).toHaveBeenCalled();
            });

            for (const [newerVersion, expectedMessage] of differentVersionTestCases) {
                it(`displays correct message when there is newer version: ${newerVersion}`, async () => {
                    // @ts-expect-error
                    getLatestReleaseVersionMock.mockResolvedValue(newerVersion);

                    await versionCheck.run(version);

                    expect(loggerMock.info.mock.calls).toMatchSnapshot();
                });
            }

            it("displays correct message when the installed version not meeting project", async () => {
                // @ts-expect-error
                getLatestReleaseVersionMock.mockResolvedValue("0.9.23");

                await versionCheck.run("0.9.23");

                expect(loggerMock.warn.mock.calls).toMatchSnapshot();
            });
        });

        describe("subsequent run through", () => {

            const lastRunThrough = "2024-01-01T00:00:00.000Z";

            let currentTime: Date;

            beforeEach(() => {
                jest.resetAllMocks();

                existsSyncSpy.mockReturnValue(true);

                versionCheck = VersionCheck.create({
                    dataDirectory: dataDir,
                    logger: loggerMock,
                    config: projectConfig,
                    runThresholdDays
                });

                // @ts-expect-error
                getLatestReleaseVersionMock.mockResolvedValue(version);

                readFileSyncSpy.mockReturnValue(
                    Buffer.from(
                        lastRunThrough
                    )
                );

                const dateNowMock = jest.fn();
                dateNowMock.mockImplementation(() => currentTime.getTime());

                // @ts-expect-error
                Date.now = dateNowMock;

                delete process.env.CHS_DEV_NO_PROJECT_VERSION_MISMATCH_WARNING;

                writeFileSyncSpy.mockImplementation((_, __) => { });

            });

            it("checks version when after time passed", async () => {
                currentTime = new Date(2024, 0, 16, 0, 0, 0, 0);

                // @ts-expect-error
                timeWithinThresholdMock.mockReturnValue(false);

                // @ts-expect-error
                getLatestReleaseVersionMock.mockResolvedValue(version);

                await versionCheck.run(version);

                expect(getLatestReleaseVersionMock).toHaveBeenCalled();
            });

            it("writes out current time when has checked version", async () => {
                currentTime = new Date(2024, 0, 16, 0, 0, 0, 0);

                // @ts-expect-error
                timeWithinThresholdMock.mockReturnValue(false);

                // @ts-expect-error
                getLatestReleaseVersionMock.mockResolvedValue(version);

                await versionCheck.run(version);

                expect(writeFileSyncSpy).toHaveBeenCalledWith(
                    join(dataDir, "last-version-run-time"),
                    "2024-01-16T00:00:00.000Z"
                );
            });

            it("does not check version when time passed not passed", async () => {
                // @ts-expect-error
                timeWithinThresholdMock.mockReturnValue(true);

                // @ts-expect-error
                getLatestReleaseVersionMock.mockResolvedValue(version);

                await versionCheck.run(version);

                expect(getLatestReleaseVersionMock).not.toHaveBeenCalled();
            });

            it("checks version when CHS_DEV_CHECK_VERSION is set", async () => {
                process.env.CHS_DEV_CHECK_VERSION = "1";

                currentTime = new Date(2024, 0, 13, 0, 0, 0, 0);

                // @ts-expect-error
                getLatestReleaseVersionMock.mockResolvedValue(version);

                await versionCheck.run(version);

                expect(getLatestReleaseVersionMock).toHaveBeenCalled();
            });

            it("displays correct message when the installed version not meeting project", async () => {
            // @ts-expect-error
                getLatestReleaseVersionMock.mockResolvedValue("0.9.23");

                await versionCheck.run("0.9.23");

                expect(loggerMock.warn.mock.calls).toMatchSnapshot();
            });

            it("does not display message when the installed version not meeting project and env var set", async () => {
            // @ts-expect-error
                getLatestReleaseVersionMock.mockResolvedValue("0.9.23");

                process.env.CHS_DEV_NO_PROJECT_VERSION_MISMATCH_WARNING = "true";

                await versionCheck.run("0.9.23");

                expect(loggerMock.warn).not.toHaveBeenCalled();
            });

        });

    });

});
