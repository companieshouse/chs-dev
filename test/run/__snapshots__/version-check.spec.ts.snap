// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VersionCheck run first run through displays correct message when the installed version not meeting project 1`] = `
[
  [
    "WARNING: this version: 0.9.23 does not meet the project requirements: >=1.0.0 <2.0.0

Run:

	chs-dev sync

to update chs-dev to a suitable version
--------------------------------------------------------------------------------

",
  ],
]
`;

exports[`VersionCheck run first run through displays correct message when there is newer version: 1.1.2 1`] = `
[
  [
    "================================================================================

ðŸ“£ There is a newer version (1.1.2) available (current version: 1.1.1)

Run:

	chs-dev sync

to update to latest version


================================================================================",
  ],
]
`;

exports[`VersionCheck run first run through displays correct message when there is newer version: 1.2.0 1`] = `
[
  [
    "================================================================================

ðŸ“£ There is a newer minor version (1.2.0) available (current version: 1.1.1)

Run:

	chs-dev sync

to update to latest version


================================================================================",
  ],
]
`;

exports[`VersionCheck run first run through displays correct message when there is newer version: 1.2.1 1`] = `
[
  [
    "================================================================================

ðŸ“£ There is a newer minor version (1.2.1) available (current version: 1.1.1)

Run:

	chs-dev sync

to update to latest version


================================================================================",
  ],
]
`;

exports[`VersionCheck run first run through displays correct message when there is newer version: 2.0.0 1`] = `
[
  [
    "================================================================================

ðŸ“£ There is a newer major version (2.0.0) available (current version: 1.1.1)

Run:

	chs-dev sync

to update to latest version


================================================================================",
  ],
]
`;

exports[`VersionCheck run first run through displays correct message when there is newer version: 2.0.1 1`] = `
[
  [
    "================================================================================

ðŸ“£ There is a newer major version (2.0.1) available (current version: 1.1.1)

Run:

	chs-dev sync

to update to latest version


================================================================================",
  ],
]
`;

exports[`VersionCheck run first run through displays correct message when there is newer version: 2.1.0 1`] = `
[
  [
    "================================================================================

ðŸ“£ There is a newer major version (2.1.0) available (current version: 1.1.1)

Run:

	chs-dev sync

to update to latest version


================================================================================",
  ],
]
`;

exports[`VersionCheck run subsequent run through displays correct message when the installed version not meeting project 1`] = `
[
  [
    "WARNING: this version: 0.9.23 does not meet the project requirements: >=1.0.0 <2.0.0

Run:

	chs-dev sync

to update chs-dev to a suitable version
--------------------------------------------------------------------------------

",
  ],
]
`;
